src/components/sketches/Scene5_happy.ts

```diff
@@ -0,0 +1,290 @@
+  FIREWORK_LIFESPAN,
+  LIFESPAN_DECREMENT,
+  VELOCITY_CHANGE,
+  TRAIL_SIZE,
+} from '../../utils/constants';
+
+export default function sketch(p: p5): void {
+  let canvas: p5.Renderer;
+  const fireworks: Firework[] = [];
+  const ringParticles: ringParticle[] = [];
+  let gravity: p5.Vector;
+
+  p.setup = () => {
+    canvas = p.createCanvas(window.innerWidth, window.innerHeight);
+    //p.colorMode(p.HSB);
+    canvas.id('p5-background');
+    gravity = p.createVector(0, 0.095);
+    p.stroke(255);
+    p.strokeWeight(4);
+    p.background(0);
+
+    // Setting mode to degrees for spawning ringParticles in circle
+    p.angleMode(p.DEGREES);
+    p.background(0);
+    // Sets 0,0 to centre
+    p.translate(p.width / 2, p.height / 2);
+    // Max number of ringParticles
+    p.frameRate(30);
+  };
+
+  p.windowResized = () => {
+    p.resizeCanvas(window.innerWidth, window.innerHeight, true);
+  };
+
+  p.draw = () => {
+    p.background(0, 0, 0, 25); // easy trails but deletes rings quickly
+
+    p.ellipse(p.mouseX,p.mouseY, 5,5);
+    if (p.random(1) < 0.03) {
+      fireworks.push(new Firework());
+    }
+
+    for (let i = fireworks.length - 1; i >= 0; i--) {
+      fireworks[i].update();
+      fireworks[i].show();
+
+      if (fireworks[i].done()) {
+        fireworks.splice(i, 1);
+      }
+    }
+
+    // If too many ringParticles on screen
+    if (ringParticles.length > MAX_RING_PARTICLES) {
+      // Delete old ringParticles (from beginning of array)
+      ringParticles.splice(0, p.floor((ringParticles.length - MAX_RING_PARTICLES * 0.8) * EASING));
+    }
+
+    // Loop through the array and show each particle
+    for (let i = 0; i < ringParticles.length; i++) {
+      ringParticles[i].update();
+      if (ringParticles[i].delete()) {
+        ringParticles.splice(i, 1);
+      } else {
+        ringParticles[i].show();
+      }
+    }
+  };
+  // Spawn ringParticles in ring with centre mouse x and y coords
+  p.mouseClicked = () => {
+    for (let i = 0; i < RING_COVERAGE; i += 4) {
+      const c = new ringParticle(p.mouseX, p.mouseY, i);
+      ringParticles.push(c);
+    }
+  };
+
+  class Firework {
+    explodeSize: number;
+    firework: Particle;
+    exploded: boolean;
+    particles: Array<Particle>;
+
+    constructor() {
+      this.explodeSize = p.random(0, EXPLODE_RANGE);
+      this.firework = new Particle(p.random(p.width), p.height, true);
+      this.exploded = false;
+      this.particles = [];
+    }
+
+    update() {
+      if (!this.exploded) {
+        this.firework.applyForce(gravity);
+        this.firework.update();
+
+        if (this.firework.vel.y >= 0) {
+          this.exploded = true;
+          this.explode();
+        }
+      }
+
+      for (let i = this.particles.length - 1; i >= 0; i--) {
+        this.particles[i].applyForce(gravity);
+        this.particles[i].update();
+
+        if (this.particles[i].done()) {
+          this.particles.splice(i, 1);
+        }
+      }
```

Comment:
splice is an O(n) function.. meaning this could potentially be O(n^2)

Mayhaps something like the following could work
```suggestion
          this.particles.splice(i, 1);
```
```suggestion
      this.particles = this.particles.filter((particle) => {
        particle.applyForce(gravity);
        particle.update();
        return !particle.done();
      });
```

Reply:
Here are [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) explaining `Array.filter`

To give a brief explanation, `filter` returns a new array by looping through the original array. If the callback function processed on an individual element of that original array returns true, then we keep the element in the new array.

Reply:
ohhh yeah that makes sense

Reply:
great suggestion!